test/feature/dynamic_types/DynamicTypesTests.cpp

```diff
@@ -0,0 +1,7669 @@
+    md = traits<MemberDescriptor>::make_shared();
+    md->id(7);
+    md->name("bool");
+    md->type(factory->get_primitive_type(eprosima::fastdds::dds::TK_BOOLEAN));
+    ASSERT_EQ(RETCODE_OK, struct_type_builder->add_member(md));
+
+    struct_type_builder->get_all_members(members_by_id);
+    ASSERT_EQ(3, members_by_id.size());
+
+    MemberDescriptor::_ref_type tmp = traits<MemberDescriptor>::make_shared();
+    auto dm = members_by_id[3];
+    ASSERT_EQ(RETCODE_OK, dm->get_descriptor(tmp));
+    EXPECT_EQ(tmp->index(), 0u);
+    EXPECT_EQ(tmp->name(), md1->name());
+    EXPECT_EQ(tmp->type(), md1->type());
+
+    dm = members_by_id[7];
+    ASSERT_EQ(RETCODE_OK, dm->get_descriptor(tmp));
+    EXPECT_EQ(tmp->name(), md->name());
+    EXPECT_EQ(tmp->type(), md->type());
+
+    dm = members_by_id[1];
+    ASSERT_EQ(RETCODE_OK, dm->get_descriptor(tmp));
+    EXPECT_EQ(tmp->index(), 1u);
+    EXPECT_EQ(tmp->name(), md2->name());
+    EXPECT_EQ(tmp->type(), md2->type());
+
+
+    // • checking adding duplicates
+    {
+        eprosima::fastdds::dds::Log::ScopeLogs _("disable");
+        //    + duplicate name
+        md = traits<MemberDescriptor>::make_shared();
+        md->id(1);
+        md->name("int32");
+        md->type(factory->get_primitive_type(eprosima::fastdds::dds::TK_INT32));
+        EXPECT_NE(RETCODE_OK, struct_type_builder->add_member(md));
+
+        //    + duplicate id
+        md = traits<MemberDescriptor>::make_shared();
+        md->id(7);
+        md->name("dup_bool");
+        md->type(factory->get_primitive_type(eprosima::fastdds::dds::TK_BOOLEAN));
+        EXPECT_NE(RETCODE_OK, struct_type_builder->add_member(md));
+    }
+}
+
+TEST_F(DynamicTypesTests, DynamicTypeBuilderFactory)
+{
+    DynamicTypeBuilderFactory::_ref_type factory {DynamicTypeBuilderFactory::get_instance()};
+
+    // Try to create with invalid values
+    // • strings
+    DynamicTypeBuilder::_ref_type created_builder {factory->create_string_type(static_cast<uint32_t>(LENGTH_UNLIMITED))};
+    ASSERT_TRUE(created_builder);
+
+    DynamicType::_ref_type type {created_builder->build()};
+    ASSERT_TRUE(type);
+    DynamicType::_ref_type type2 {created_builder->build()};
+    ASSERT_TRUE(type2);
+
+    ASSERT_TRUE(type->equals(type2));
+
+    // • wstrings
+    created_builder = factory->create_wstring_type(static_cast<uint32_t>(LENGTH_UNLIMITED));
+    ASSERT_TRUE(created_builder);
+
+    type = created_builder->build();
+    ASSERT_TRUE(type);
+    type2 = created_builder->build();
+    ASSERT_TRUE(type2);
+
+    ASSERT_TRUE(type->equals(type2));
+}
+
+TEST_F(DynamicTypesTests, DynamicType_int32)
+{
+    DynamicTypeBuilderFactory::_ref_type factory {DynamicTypeBuilderFactory::get_instance()};
+
+    DynamicType::_ref_type created_type {factory->get_primitive_type(eprosima::fastdds::dds::TK_INT32)};
+    ASSERT_TRUE(created_type);
+    DynamicData::_ref_type data {DynamicDataFactory::get_instance()->create_data(created_type)};
```

Comment:
DynamicData non-tested API:

* `type`
* `get_descriptor`
* `get_int8_value`/`set_int8_value`/`get_int8_values`
* `get_uint8_values`/`set_uint8_values`
* `get_int16_values`/`set_int16_values`
* `get_uint16_values`/`set_uint16_values`
* `get_int64_values`/`set_int64_values`
* `get_uint64_values`/`set_uint64_values`
* `get_float32_values`/`set_float32_values`
* `get_float64_values`/`set_float64_values`
* `get_float128_values`/`set_float128_values`
* `get_char8_values`/`set_char8_values`
* `get_char16_values`/`set_char16_values`
* `get_byte_values`/`set_byte_values`
* `get_boolean_values`/`set_boolean_values`
* `get_string_values`/`set_string_values`
* `get_wstring_values`/`set_wstring_values`

Most of this API is used and tested in the dds types tests, however, the corresponding negative tests should be included in this PR.